test/differentials.jl

```diff
@@ -49,6 +49,25 @@
         @test conj(o) == o
     end
 
+	@testset "Thunk" begin
+		@test @thunk(3) isa Thunk
+
+		@testset "show" begin
+			rep = repr(Thunk(rand))
+			@test occursin(r"Thunk\(.*rand.*\)", rep)
+		end
+
+		@testset "Externing" begin
+			@test extern(@thunk(3)) == 3
+			@test extern(@thunk(@thunk(3))) == 3
+		end
+
+		@testset "calling thunks should call inner function" begin
+			@test (@thunk(3))() == 3
+			@test (@thunk(@thunk(3)))() isa Thunk
```

Comment:
Why isn't thunking a thunk a no-op?

Reply:
becausse detecting that is really hard. Probably impossible in the general case.
Remember that
`@thunk(x)` translates into `Thunk(()->x)`
So 
`@thunk(@thunk(x))` turns into `Thunk(()->Thunk(()->x))`
*Not* into `Thunk(Thunk(x))`

so if I had something like
```
f(x) = rand() > 0.5 ? x : Thunk(()->svd(x).U)
g(x) = @thunk(f(x))
```

Can't really deal with it

Reply:
I see. Unfortunate but Makes sense. Thanks

Reply:
If we only had the `@thunk` macro, not the function, could thunking a closure be a no-op? Would that help us avoid complexity from nested thunks?

Reply:
I am not sure,
if that is possibke it would have tio happen with a runtime check
every time `@thunk` is used,
which probably introduces worse performance than this change,
which so far seems to compile out anyway.